### DISCUSSION TITLE
Listen error in torrent library on some devices

### DISCUSSION THREAD

**Discussion by tsynik** [anacrolix/torrent#856] [General]
@anacrolix, we use anacrolix@torrent as library in TorrServer project. After last library changes and added holepunching support, here is error on sockets open with linux binary running on old Android 4.3 device:

BTS.Connect() error! first listen: listen tcp4 :32000: protocol not available

TorrServer code: https://github.com/YouROK/TorrServer/blob/b355c95296cb987747d24c0244ff3f626edc2ca6/server/web/server.go#L37-L41
torrent library part: https://github.com/anacrolix/torrent/blob/b730a99a480d342f128f84f9c45201133e3c7e11/socket.go#L121-L123

so it fails on listenAll and listenAllRetry after. What can cause such error? port 32000 not used according netstat. And there is no such error on other Devices with android binary. Older anacrolix@torrent worked on this TV.

p.s. last working library version on this device is torrent v1.50.0. it doesn't work since torrent v1.51.0 release

Probably those TorrServer issues also related: https://github.com/YouROK/TorrServer/issues/269 https://github.com/YouROK/TorrServer/issues/270

**Comment by anacrolix** [comment]
Thanks for the report. I'll check to see if there's anything I could have changed.

**Comment by anacrolix** [comment]
Could it be this:

https://github.com/anacrolix/torrent/commit/599846546ec6e34036de889a4f41ea3f579153a3#diff-5bbc0182584739004c6ad9b1479cfd3dd1c31e43db4bf8a51da3045f4ff2228dR80

And probably this subsequent call:
https://github.com/anacrolix/torrent/commit/599846546ec6e34036de889a4f41ea3f579153a3#diff-da26e0f35e0d59c26d6d21dc1ea9c8c7b365b3865606da8afe8366eb2a668597R23

It might not work on your Linux for some reason. Since I removed TCP dialling out from specific addresses, it's possible this is no longer needed. It can at least just be a log message, or ignored completely if it fails.

**Comment by tsynik** [comment]
> Could it be this:

@anacrolix builded TorrServer with torrent v1.52.4 and mentioned lines commented - there is no error:
2023/08/10 10:17:00 Start TorrServer MatriX.125 torrent v1.52.4
...
2023/08/10 10:17:00 Start web server at port 8090
...
and all is OK

**Reply by anacrolix** [reply]
Thanks. Want to make a PR?

**Comment by tsynik** [comment]
> Thanks. Want to make a PR?

I'm not sure commenting those 2 lines is a proper way to fix this issue, because it will disable REUSEPORT for all linuxes.

**Reply by anacrolix** [reply]
Yeah, sorry I didn't explain, but I don't think I ended up needing reuseport due to removing the change that outgoing TCP would dial from the listening port. I'll make the change and do some tests.

**Reply by anacrolix** [reply]
@tsynik could you try this branch? https://github.com/anacrolix/torrent/tree/no-reuseport

`go get github.com/anacrolix/torrent@no-reuseport`. If it works for you I'll make it the next release. Thanks!

**Comment by tsynik** [comment]
> `go get github.com/anacrolix/torrent@no-reuseport`. If it works for you I'll make it the next release. Thanks!

@anacrolix hm, I have the same error with this branch:
```
TorrServer MatriX.125, go1.21.0, CPU Num: 4
...
2023/08/14 10:35:23 Check web port 8090
2023/08/14 10:35:23 Start TorrServer MatriX.125 torrent v1.52.5-0.20230814013847-cc4a8a4f9e20
2023/08/14 10:35:23 Local IPs: [192.168.77.55]
2023/08/14 10:35:23 Set listen port to random autoselect (0)
2023/08/14 10:35:23 Client config: {"CacheSize":67108864,"ReaderReadAHead":95,"PreloadCache":50,"UseDisk":false,"TorrentsSavePath":"","RemoveCacheOnDrop":false,"ForceEncrypt":false,"RetrackersMode":1,"TorrentDisconnectTimeout":30,"EnableDebug":false,"EnableDLNA":false,"FriendlyName":"","EnableRutorSearch":false,"EnableIPv6":false,"DisableTCP":false,"DisableUTP":false,"DisableUPNP":false,"DisableDHT":false,"DisablePEX":false,"DisableUpload":false,"DownloadRateLimit":0,"UploadRateLimit":0,"ConnectionsLimit":25,"PeersListenPort":0}
2023/08/14 10:35:23 BTS.Connect() error! first listen: listen tcp4 :0: protocol not available
```
probably because of this config: https://github.com/anacrolix/torrent/blob/cc4a8a4f9e20e91f18896c56696e8fe0a2129916/socket.go#L40-L53

but after commenting only 1 line 23 at unix socket code (https://github.com/anacrolix/torrent/blob/cc4a8a4f9e20e91f18896c56696e8fe0a2129916/sockopts_unix.go#L23) it starts without errors:
```
TorrServer MatriX.125, go1.21.0, CPU Num: 4
...
2023/08/14 10:56:18 Check web port 8090
2023/08/14 10:56:18 Start TorrServer MatriX.125 torrent v1.52.5-0.20230814013847-cc4a8a4f9e20
2023/08/14 10:56:18 Local IPs: [192.168.77.55]
2023/08/14 10:56:18 Set listen port to random autoselect (0)
2023/08/14 10:56:18 Client config: {"CacheSize":67108864,"ReaderReadAHead":95,"PreloadCache":50,"UseDisk":false,"TorrentsSavePath":"","RemoveCacheOnDrop":false,"ForceEncrypt":false,"RetrackersMode":1,"TorrentDisconnectTimeout":30,"EnableDebug":false,"EnableDLNA":false,"FriendlyName":"","EnableRutorSearch":false,"EnableIPv6":false,"DisableTCP":false,"DisableUTP":false,"DisableUPNP":false,"DisableDHT":false,"DisablePEX":false,"DisableUpload":false,"DownloadRateLimit":0,"UploadRateLimit":0,"ConnectionsLimit":25,"PeersListenPort":0}
2023/08/14 10:56:18 Start web server at port 8090
```

**Reply by anacrolix** [reply]
Woops! Thanks for the catch. I've pushed this fix to the same branch.

**Comment by tsynik** [comment]
> Woops! Thanks for the catch. I've pushed this fix to the same branch.

now it's OK without any local changes:
```
TorrServer MatriX.125, go1.21.0, CPU Num: 4
...
2023/08/14 18:16:18 Check web port 8090
2023/08/14 18:16:18 Start TorrServer MatriX.125 torrent v1.52.5-0.20230814125102-43fe6476c693
2023/08/14 18:16:18 Local IPs: [192.168.77.55]
2023/08/14 18:16:18 Set listen port to random autoselect (0)
2023/08/14 18:16:18 Client config: {"CacheSize":67108864,"ReaderReadAHead":95,"PreloadCache":50,"UseDisk":false,"TorrentsSavePath":"","RemoveCacheOnDrop":false,"ForceEncrypt":false,"RetrackersMode":1,"TorrentDisconnectTimeout":30,"EnableDebug":false,"EnableDLNA":false,"FriendlyName":"","EnableRutorSearch":false,"EnableIPv6":false,"DisableTCP":false,"DisableUTP":false,"DisableUPNP":false,"DisableDHT":false,"DisablePEX":false,"DisableUpload":false,"DownloadRateLimit":0,"UploadRateLimit":0,"ConnectionsLimit":25,"PeersListenPort":0}
...
2023/08/14 18:16:18 Start web server at port 8090
```

**Reply by anacrolix** [reply]
Thanks for seeing this through. I'll merge and release it now.

**Reply by anacrolix** [reply]
v1.52.5
